### ISSUE TITLE
Bug with property serializers

### ISSUE THREAD

**Issue by ProGM** [neo4jrb/neo4j#1128]
I was trying to implement enums #1082 , and I discovered a bug with `serializers`.

Here's a test that demonstrates it:

```ruby
describe Neo4j::ActiveNode do
  before(:each) do
    # This serializer adds a text when the data is saved on the db,
    # and removes it when deserializing
    class MySerializer
      def initialize(text)
        @text = text
      end

      def converted?(value)
        value.is_a?(db_type)
      end

      def db_type
        String
      end

      def convert_type
        Symbol
      end

      def to_ruby(value)
        value.gsub(@text, '').to_sym if value
      end

      alias_method :call, :to_ruby

      def to_db(value)
        "#{value}#{@text}" if value
      end
    end

    stub_active_node_class('MyData') do
      property :polite_string, serializer: MySerializer.new(', sir.')
      property :happy_string, serializer: MySerializer.new(' :)')
    end
  end

  it 'serializes correctly' do
    data = MyData.new
    data.polite_string = :hello
    data.happy_string = :hello
    data.save!
    expect(MyData.as(:d).pluck('d.polite_string, d.happy_string')).to eq([['hello, sir.', 'hello :)']])
  end
end
```

Here's the test output:

```ruby
1) Neo4j::ActiveNod serializes correctly
     Failure/Error: expect(MyData.as(:d).pluck('d.polite_string, d.happy_string')).to eq([['hello, sir.', 'hello :)']])

       expected: [["hello, sir.", "hello :)"]]
            got: [["hello :)", "hello :)"]]

       (compared using ==)
     # ./spec/e2e/enum_spec.rb:50:in `block (2 levels) in <top (required)>'
```

EDIT:
After exploring the code, I noticed that the `convert_type` output is used as a key to determine what is the correct TypeConverter for a certain type, and it's used for ALL properties.
https://github.com/neo4jrb/neo4j/blob/77190ab416d219f230dc86eb71e3b7b92df65b0b/lib/neo4j/shared/declared_property.rb#L115
It's a bit strange, since I'm passing the `serializer` setting only to a specific property, not as a global setting.

My question is: it's really a bug? or a design choice?

I'm a bit confused.

**Comment by cheerfulstoic** [comment]
I'm not very familiar with serializers.  @subvertallchris is this an ActiveAttr thing, or something we put in?

**Comment by subvertallchris** [comment]
It's ours, I will look over this today.

**Comment by subvertallchris** [comment]
Yesterday was a bit busy so I'm finally taking a look now. I'm pretty sure that the `serializer` option (originally called `type_converter`) came from someone who was contributing for a short time. I don't have time right this minute to figure out how it's supposed to work -- it seems a little vague to me and we have other ways of accomplishing this -- but I found another way to get your test passing. It might not be _the_ solution but I think it might point us in the right direction.

Check out #1134. I fixed the `serialize` class method to get your test passing and allow us to define serializers that do not dominate an entire primitive type. If we want to keep using the `serialize` property option, we can just have it delegate to `serializer` so they can share that code and behavior, just as we do for the `index` option and class methods.
